linker.ts

```diff
@@ -1,72 +1,133 @@
+/**
+ * Generates a new-style library placeholder from a fully-qualified library name.
+ *
+ * Newer versions of the compiler use hashed names instead of just truncating the name
+ * before putting it in a placeholder.
+ *
+ * @param fullyQualifiedLibraryName
+ */
+function libraryHashPlaceholder (fullyQualifiedLibraryName) {
+  return `$${keccak256(fullyQualifiedLibraryName).slice(0, 34)}$`;
 }
 
-const linkBytecode = function (bytecode, libraries) {
+/**
+ * Used to support legacy library names and the current hash library names,
+ * replaces the names with the provided hex address.
+ *
+ * @param bytecode
+ * @param address
+ * @param label
+ */
+function replacePlaceholder (bytecode, label, address) {
+  // truncate to 36 characters
+  const truncatedName = label.slice(0, 36);
+  const libLabel = `__${truncatedName.padEnd(36, '_')}__`;
+
+  while (bytecode.indexOf(libLabel) >= 0) {
+    bytecode = bytecode.replace(libLabel, address);
+  }
+
+  return bytecode;
+}
+
+/**
+ * Finds and all library placeholders in the provided bytecode and replaces them with actual addresses.
+ * Supports both old- and new-style placeholders (even both in the same file).
+ * See [Library Linking](https://docs.soliditylang.org/en/latest/using-the-compiler.html#library-linking)
+ * for a full explanation of the linking process.
+ *
+ * Example of a legacy placeholder: `__lib.sol:L_____________________________`
+ * Example of a new-style placeholder: `__$cb901161e812ceb78cfe30ca65050c4337$__`
+ *
+ * @param bytecode Hex-encoded bytecode string. All 40-byte substrings starting and ending with
+ *     `__` will be interpreted as placeholders.
+ *
+ * @param libraries Mapping between fully qualified library names and the hex-encoded
+ *     addresses they should be replaced with. Addresses shorter than 40 characters are automatically padded with zeros.
+ *
+ * @returns bytecode Hex-encoded bytecode string with placeholders replaced with addresses.
+ *    Note that some placeholders may remain in the bytecode if `libraries` does not provide addresses for all of them.
+ */
+function linkBytecode (bytecode, libraries) {
   assert(typeof bytecode === 'string');
   assert(typeof libraries === 'object');
+
   // NOTE: for backwards compatibility support old compiler which didn't use file names
   const librariesComplete = {};
-  for (const libraryName in libraries) {
-    if (typeof libraries[libraryName] === 'object') {
-      // API compatible with the standard JSON i/o
-      for (const lib in libraries[libraryName]) {
-        librariesComplete[lib] = libraries[libraryName][lib];
-        librariesComplete[libraryName + ':' + lib] = libraries[libraryName][lib];
-      }
-    } else {
-      // backwards compatible API for early solc-js versions
-      const parsed = libraryName.match(/^([^:]+):(.+)$/);
-      if (parsed) {
-        librariesComplete[parsed[2]] = libraries[libraryName];
+
+  for (const [libraryName, library] of Object.entries(libraries)) {
```

Comment:
Or we could adjust it here:
```suggestion
  for (const [fullyQualifiedLibraryName, address] of Object.entries(libraries)) {
```